Add removed users page

AlexVelezLl opened the pull request:
## Summary

* Adds removed users page.
* ~This branch was created out of the #13600 branch, to not duplicating code and avoid merge conflicts. So do not merge until #13600 is merged.~

https://github.com/user-attachments/assets/3a49a2f5-f2e1-47e6-90fa-f0db50448439


## References
Closes #13386

## Reviewer guidance
* Go to facility > users table and remove multiple users.
* Go to options > removed users.
* Check that all removed users appears on the list.
* Play with the recover/permanent delete actions.
* Check that removed users cannot be edited in any other way.

## Notes
* Didn't add the image that is shown in the empty table state yet, as it is not clear yet what does that image does.
* I made an assumption here: I found some comments in the figma, and it seems the intention in the removed users table was to show the remaining days before a user is permanently deleted, instead of how many days ago a user was soft deleted. And highlighting when there were just 7 days left before permanent deletion was making more sense to me rather to highlight that the user was soft-deleted <=7 days ago like in the figma specs. Also, since the notice on top of the table in figma says "Records will show the days remaining before..." I assumed it was "before permanent deletion".

AlexVelezLl committed:
add initial users trash page

AlexVelezLl committed:
Show deletion date

AlexVelezLl committed:
Add permanent delete and recover actions

LianaHarris360 commented on packages/kolibri-common/strings/bulkUserManagementStrings.js:
```diff
@@ -418,4 +447,61 @@ export const bulkUserManagementStrings = createTranslator('BulkUserManagementStr
+    context: 'Label for the button to recover multiple users from the trash.',
+  },
+  permanentDeletion: {
+    message: 'Permanent deletion',
```
In the Figma designs, this column is labeled "Deleted." I know the strings might eventually change, but this also impacts how the date is shown in this column. Originally, it's supposed to display how long ago a user was deleted, for example, "20 days ago," but now it shows a future time, such as "next month." Was there a specific reason for this change?

AlexVelezLl replied:
Oh, apologies, just noticed I should have mentioned this assumption in the PR description.

I thought the designs were a bit stale, I found this comment in the figma, and it seems the intention was to show the remaining days instead. And highlighting when there were just 7 days left before permanent deletion was making more sense to me rather to highlight that the user was soft-deleted <=7 days ago like in the figma specs. Also because of the notice on top of the table that says "Records will show the days remaining before..." I assumed it was before permanent deletion :sweat_smile:. So that's why I framed this as days before permanent deletion, rather than soft deleted X days ago, we can chat more with @tomiwaoLE what was the intention here.

<img width="625" height="918" alt="image" src="https://github.com/user-attachments/assets/8fb2e35c-9e1e-447e-bb2a-59af9cf21460" />

LianaHarris360 commented on kolibri/plugins/facility/assets/src/views/users/common/UsersTable.vue:
```diff
@@ -482,6 +541,10 @@
         // If logged-in user is a superuser, then they can edit anybody (including other SUs).
         // Otherwise, only non-SUs can be edited.
+        // If showing deleted users, they cannot be edited.
+        if (isShowingDeletedUsers.value) {
```
Instead of disabling the options menu button completely and leaving the last column as dead space, maybe the only available action should be to open the `PermanentDeleteModal` to permanently delete the user.

AlexVelezLl replied:
Sounds good, will do!

LianaHarris360 reviewed (commented):
Overall, these changes align with the Figma designs. I have a question about the column label change from "Deleted" to "Permanent deletion" and noted a suggestion regarding the menu button in the last column on the removed users page.

AlexVelezLl committed:
Populate user dropdown options

AlexVelezLl commented:
Thanks @LianaHarris360! I just pushed a commit to show the options button for each user. I also added the recover action, since it made sense to me to see both actions we can do with soft deleted users, but if you have any thoughts, just let me know!

nucleogenesis commented on kolibri/plugins/facility/assets/src/views/users/UsersTrashPage/PermanentDeleteModal.vue:
```diff
@@ -0,0 +1,119 @@
+        defaultErrorMessage$,
+      } = bulkUserManagementStrings;
+
+      const close = () => {
```
(non-blocking) curious why the function rather than just using `emit('close')` throughout?

AlexVelezLl replied:
I usually like to abstract actions like this in just a method (in this case "close") and not depend on the actual implementation for the action in more than 1 place (in this case, we call close in three different places). If for some reason we need this modal to be the view of a route, and need to implement the `goBack` methods, then we would only need to update the close implementation within this method, and not in many places, or if for some reason we need to do a cleanup before closing the modal we will only need to do 1 update. Although for the specific case of this modal this is less probable :sweat_smile:.

nucleogenesis reviewed (commented):
Code looks great in general here - easy to read and follow how things are going. I'll await QA for final approval

pcenov commented:
Hi @AlexVelezLl, here's a list with the observed issues:

1. It's possible to create a user with the same username as a temporary removed user. This should be avoided as it can be problematic:

https://github.com/user-attachments/assets/eb35a274-dc9c-492f-a1af-9a19198d786a

2. After removing a user the admin should be brought back to the Users table in its default state, because the operation is completed:

https://github.com/user-attachments/assets/94d8d847-7e76-4dc0-a2fd-90f8814c97ce

3.  The same is valid for the Removed users page, where if I have searched for a specific user and deleted that user then I am seeing a "No removed users" while there are several removed users, but the filter is still being applied:

https://github.com/user-attachments/assets/5567488e-ddf4-4d62-b5b9-81e1c16272d9

4. Currently the page is called "Removed Users" while it should be "Removed users". Also I think that the text "Records will show the days remaining before permanent deletion." should be shown only when there are removed users, otherwise it's just a duplication of "Removed users are stored here for 30 days before being permanently deleted"

<img width="798" height="454" alt="texts" src="https://github.com/user-attachments/assets/c2d0dd72-3269-46b4-9e78-121ce618a686" />

5. The snackbar message after having recovered a user is incorrect:

<img width="1332" height="866" alt="NaN" src="https://github.com/user-attachments/assets/7e47dd47-003c-44ef-b6fa-5d301f8c4d03" />

6. Recovering a set of 50 users results in a "Resource not found" error shown to the user and a 404 error in the console but the users are actually recovered successfully:

https://github.com/user-attachments/assets/a1211c20-7b34-48a9-8db6-e8173ff201f6

7. When I remove a user the modal says that the user is removed from all classes and immediately deactivated, however if I am signed in as that user I can seemingly continue to use the system until I sign out. Also I am temporary getting error screens. This could be problematic in the case of coaches, admin and super admin users and should be handled better:

https://github.com/user-attachments/assets/6cf8d0ed-5b8f-4f10-9f09-a57d7321df05

8. If there are no search results matching the entered search term then I should be able to clear it while it shows the generic "No removed users" text with no option to modify the search:

https://github.com/user-attachments/assets/1dba031d-91e7-4b52-a592-8bc4598025ec

AlexVelezLl commented:
Thanks @pcenov!
Points 1, 2, 3 and 7 have not been introduced in the scope of this PR, some decisions are needed for these cc: @nucleogenesis @LianaHarris360.

Will fix 4, 5, 6 and 8!

AlexVelezLl committed:
Fix 404 pagination error on users removal, keep users table while searching

AlexVelezLl commented:
Hi @pcenov. Points 4, 5, 6 and 8 should be solved now. 8 was also happening in the "New users page", and also fixed it here. (edit: 7 should be also fixed)

6 was also happening if we remove the last user of the last page in all users table, and just fixed it too cc: @nucleogenesis @LianaHarris360.

AlexVelezLl committed:
Add empty images

nucleogenesis commented:
1) (non-blocking, will follow-up) The username validation for creation should check uniqueness w/ the `all_objects` manager. 
2) (non-blocking, will follow-up) I'm not opposed to something like "When the user has a search query AND the user's actions make the search results empty, then we should clear their search". But I lean more toward the idea of "the user's query remains until they clear it themselves".
3) **blocking** @AlexVelezLl the main thing to fix here is that the user can no longer clear their query. I think that if we keep the search/filters toolbar visible and only replace the table itself w/ the "No results" message this should be good to go.
7) _Please investigate_  I think it'd be worth taking a short bit of time @AlexVelezLl to look at what is happening in this instance (ie, is the user's session being deleted when the user is soft deleted?)

Thanks for the review @pcenov -

AlexVelezLl commented:
> blocking @AlexVelezLl the main thing to fix here is that the user can no longer clear their query. I think that if we keep the search/filters toolbar visible and only replace the table itself w/ the "No results" message this should be good to go.

That is already fixed @nucleogenesis :sweat_smile: as its the same point as 8

AlexVelezLl committed:
Add invalidate removed users session

pcenov reviewed (approved):
Thanks @AlexVelezLl - I confirm that 4, 5, 6 and 8 are fixed now and I didn't spot anything new while retesting today.
